ibis/backends/dask/tests/test_udf.py

```diff
@@ -0,0 +1,483 @@
+        {'key': list('ab'), 'my_add': [sum([1.0 + 4.0, 2.0 + 5.0]), 3.0 + 6.0]}
+    )
+    tm.assert_frame_equal(result, expected)
+
+
+def test_udaf(con, t, df):
+    expr = my_string_length_sum(t.a)
+
+    assert isinstance(expr, ir.ScalarExpr)
+
+    result = expr.execute()
+    expected = t.a.execute().str.len().mul(2).sum()
+    assert result == expected
+
+
+def test_udaf_analytic(con, t, df):
+    expr = zscore(t.c)
+
+    assert isinstance(expr, ir.ColumnExpr)
+
+    result = expr.execute()
+
+    def f(s):
+        return s.sub(s.mean()).div(s.std())
+
+    expected = (f(df.c)).compute()
+    tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.xfail(
+    raises=NotImplementedError, reason='TODO - windowing - #2553'
+)
+def test_udaf_analytic_groupby(con, t, df):
+    expr = zscore(t.c).over(ibis.window(group_by=t.key))
+
+    assert isinstance(expr, ir.ColumnExpr)
+
+    result = expr.execute()
+
+    def f(s):
+        return s.sub(s.mean()).div(s.std())
+
+    expected = df.groupby('key').c.transform(f)
+    tm.assert_series_equal(result, expected)
+
+
+def test_udaf_groupby(df3):
+    con = connect({'df3': df3})
+    t = con.table('df3')
+
+    expr = t.groupby(t.key).aggregate(my_corr=my_corr(t.a, t.b))
+
+    assert isinstance(expr, ir.TableExpr)
+
+    result = expr.execute().sort_values('key').reset_index()
+
+    dfi = df3.set_index('key').compute()
+    expected = pd.DataFrame(
+        {
+            'key': list('def'),
+            'my_corr': [
+                dfi.loc[value, 'a'].corr(dfi.loc[value, 'b'])
+                for value in 'def'
+            ],
+        }
+    )
+
+    columns = ['key', 'my_corr']
+    tm.assert_frame_equal(result[columns], expected[columns])
+
+
+def test_nullable():
+    t = ibis.table([('a', 'int64')])
+    assert nullable(t.a.type()) == (type(None),)
+
+
+def test_nullable_non_nullable_field():
+    t = ibis.table([('a', dt.String(nullable=False))])
+    assert nullable(t.a.type()) == ()
+
+
+def test_udaf_parameter_mismatch():
+    with pytest.raises(TypeError):
+
+        @reduction(input_type=[dt.double], output_type=dt.double)
+        def my_corr(lhs, rhs, **kwargs):
+            pass
+
+
+def test_udf_parameter_mismatch():
+    with pytest.raises(TypeError):
+
+        @reduction(input_type=[], output_type=dt.double)
+        def my_corr2(lhs, **kwargs):
+            pass
+
+
+def test_udf_error(t):
+    @elementwise(input_type=[dt.double], output_type=dt.double)
+    def error_udf(s):
+        raise ValueError('xxx')
+
+    with pytest.raises(ValueError):
+        error_udf(t.c).execute()
+
+
+def test_compose_udfs(t2, df2):
+    expr = times_two(add_one(t2.a))
+    result = expr.execute().reset_index(drop=True)
+    expected = df2.a.add(1.0).mul(2.0).compute()
+    tm.assert_series_equal(result, expected, check_names=False)
+
+
+@pytest.mark.xfail(
+    raises=NotImplementedError, reason='TODO - windowing - #2553'
+)
+def test_udaf_window(t2, df2):
```

Comment:
It did feel like some of the tests here are similar to the test in https://github.com/ibis-project/ibis/blob/master/ibis/backends/tests/test_window.py#L375 for example

I'd suggest sharing as much of these as possible to the backends/tests so other backends can benefit from the same test.

Reply:
All these tests were lifted directly from the pandas backend which makes tracking feature parity a lot easier. I'm happy to further refactor those out of both backends, but I'd like to do that separately.

Reply:
Yeah that's fine. Ideally the more shared tests the better.